Fixing startup and tests for H2 2.x

ahus1 opened the pull request:
This is a draft PR to show what is necessary to make Keycloak start (works in Untertow mode) and fix the tests (still pending).

ahus1 commented:
@stianst / cc: @hmlnarik - this is a first set of changes. 

TL;DR: Upgrade still seems feasible, more work needs to be done. Unfortunately our Hibernate version doesn't support H2 2.x, so I try to run H2 2.x in LEGACY mode and hope for the best. 

The following was needed to start up the server using undertow - Quarkus will most likely still fail: 

* `VALUE` is now a keyword and needs to be quoted (even when running H2 in legacy mode) -> some changesets needed to be changed. Will apply later any `<validCheckSum/>` entries as needed
* Hibernate would need to upgrade from 5.3.24 to  5.6.5. Until then, adding `;MODE=LEGACY` to the JDBC URL for H2 databases, as this would otherwise fail for example when inserting booleans into the database
* Hibernate need to quote `VALUE` for column name - so I added an update H2Dialect that's auto-detected by `DefaultJpaConnectionProviderFactory`. Seems to be still a problem with latest Hibernate versions. 

Will have a look next week on how to continue this. Your feedback is appreciated.

stianst commented:
@ahus1 looks like the Quarkus dist is already using Hibernate 5.6: https://github.com/keycloak/keycloak/blob/148c7695ff5b32f46f05a5f341ca5456da144987/quarkus/pom.xml#L40

ahus1 commented:
@stianst - true; updated this in the main repo and now the H2 legacy mode is no longer needed. Still needed to add a dialect to care for the "VALUE" Keyword. Added the dialect for Quarkus, now the local Quarkus startup works.

Let's see how the tests complete, and then clean up later.

ahus1 commented:
Found an JDBC URL option `;NON_KEYWORDS=VALUE` to avoid changing the existing Liquibase scripts, and also avoid changing Hibernate settings for several entities for all databases. This should keep the change minimal. Already discussed with @hmlnarik. 

Will close this PR; already found a issue in the main repo `#12607` (don want to link here, for now). Will create a PR for that issue later this week.

ahus1 commented:
PR opened in main keycloak repo: https://github.com/keycloak/keycloak/pull/14819
